File: src/clu/conversions.py
"""Doc Incomplete."""


def bytes_to_si(size: float) -> str:
    """
    Convert a number of bytes to a human-readable SI string (e.g., 1536 -> '1.5 KB').
    """
    units = ["B", "KB", "MB", "GB", "TB", "PB", "EB"]
    size = float(size)
    for unit in units:
        if size < 1024:
            return f"{size:.1f} {unit}"
        size /= 1024
    return f"{size:.1f} {units[-1]}"


def si_to_bytes(size_str: str) -> float:
    """
    Convert a human-readable SI string (e.g., '1.5 KB') to bytes (int).
    """
    units = {
        "B": 0,
        "K": 1,
        "KB": 1,
        "M": 2,
        "MB": 2,
        "G": 3,
        "GB": 3,
        "T": 4,
        "TB": 4,
        "P": 5,
        "PB": 5,
        "E": 6,
        "EB": 6,
    }
    size_str = size_str.strip().upper()
    number_str = ""
    unit = ""
    for char in size_str:
        if char.isdigit() or char == "." or char == ",":
            number_str += char
        elif char.isalpha():
            unit += char
    number = float(number_str.replace(",", ""))
    unit = unit.strip()
    if unit not in units:
        raise ValueError(f"Unknown unit: {unit}")
    exponent = units[unit]
    return float(number * (1024**exponent))


def seconds_to_text(secs: int) -> str:
    """
    Convert seconds to a human-readable string with weeks, days, hours, and minutes.

    Args:
        secs: Number of seconds to convert

    Returns:
        A formatted string like "1 week, 2 days, 3 hours, 45 minutes"
    """
    if secs < 0:
        return "0 seconds"

    # Calculate time units
    months = secs // 2592000  # 30 * 24 * 60 * 60
    remaining = secs % 2592000

    days = remaining // 86400  # 24 * 60 * 60
    remaining = remaining % 86400

    hours = remaining // 3600  # 60 * 60
    remaining = remaining % 3600

    minutes = remaining // 60
    seconds = remaining % 60

    # Build result parts
    parts = []
    if months > 0:
        parts.append(f"{months} month{'s' if months != 1 else ''}")
    if days > 0:
        parts.append(f"{days} day{'s' if days != 1 else ''}")
    if hours > 0:
        parts.append(f"{hours} hour{'s' if hours != 1 else ''}")
    if minutes > 0:
        parts.append(f"{minutes} minute{'s' if minutes != 1 else ''}")
    if seconds > 0 or len(parts) == 0:
        parts.append(f"{seconds} second{'s' if seconds != 1 else ''}")

    return ", ".join(parts)

File: pkg/conversions.go
package pkg

// Port of src/clu/conversions.py providing byte size and time formatting
// plus parsing of simple SI size strings.

import (
	"errors"
	"fmt"
	"math"
	"strings"
)

// BytesToSI converts a byte count to a human-readable SI string (base 1024) with one decimal.
// Example: 1536 -> "1.5 KB"
func BytesToSI(size float64) string {
	units := []string{"B", "KB", "MB", "GB", "TB", "PB", "EB"}
	val := size
	for _, u := range units {
		if val < 1024.0 || u == units[len(units)-1] {
			return fmt.Sprintf("%.1f %s", val, u)
		}
		val /= 1024.0
	}
	return fmt.Sprintf("%.1f %s", val, units[len(units)-1])
}

// SIToBytes parses a human-readable size string (e.g. "1.5 KB") into bytes (float64).
// Accepts units B,K,KB,M,MB,G,GB,T,TB,P,PB,E,EB (case-insensitive). Commas permitted in number.
func SIToBytes(sizeStr string) (float64, error) {
	units := map[string]int{
		"B": 0,
		"K": 1, "KB": 1,
		"M": 2, "MB": 2,
		"G": 3, "GB": 3,
		"T": 4, "TB": 4,
		"P": 5, "PB": 5,
		"E": 6, "EB": 6,
	}
	s := strings.TrimSpace(strings.ToUpper(sizeStr))
	var numberStr, unitStr string
	for _, ch := range s {
		if (ch >= '0' && ch <= '9') || ch == '.' || ch == ',' {
			numberStr += string(ch)
		} else if ch >= 'A' && ch <= 'Z' {
			unitStr += string(ch)
		}
	}
	if numberStr == "" {
		return 0, errors.New("no numeric portion found")
	}
	num, err := parseFloat(numberStr)
	if err != nil {
		return 0, err
	}
	unitStr = strings.TrimSpace(unitStr)
	exp, ok := units[unitStr]
	if !ok {
		return 0, fmt.Errorf("unknown unit: %s", unitStr)
	}
	return num * math.Pow(1024, float64(exp)), nil
}

// parseFloat handles comma removal then parsing.
func parseFloat(s string) (float64, error) {
	s = strings.ReplaceAll(s, ",", "")
	var num float64
	_, err := fmt.Sscan(s, &num)
	if err != nil {
		return 0, fmt.Errorf("invalid number: %s", s)
	}
	return num, nil
}

// SecondsToText converts seconds into a human-readable string with months, days, hours, minutes, seconds.
// Follows Python logic (using 30-day months) and pluralization rules.
func SecondsToText(secs int64) string {
	if secs <= 0 {
		return "0 seconds"
	}
	months := secs / 2592000 // 30*24*60*60
	remaining := secs % 2592000

	days := remaining / 86400
	remaining = remaining % 86400

	hours := remaining / 3600
	remaining = remaining % 3600

	minutes := remaining / 60
	seconds := remaining % 60

	parts := make([]string, 0, 5)
	if months > 0 {
		parts = append(parts, pluralize(months, "month"))
	}
	if days > 0 {
		parts = append(parts, pluralize(days, "day"))
	}
	if hours > 0 {
		parts = append(parts, pluralize(hours, "hour"))
	}
	if minutes > 0 {
		parts = append(parts, pluralize(minutes, "minute"))
	}
	if seconds > 0 || len(parts) == 0 {
		parts = append(parts, pluralize(seconds, "second"))
	}
	return strings.Join(parts, ", ")
}

func pluralize(n int64, word string) string {
	if n == 1 {
		return fmt.Sprintf("%d %s", n, word)
	}
	return fmt.Sprintf("%d %ss", n, word)
}
